subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -223,6 +233,14 @@ public void visitOutgoingDependencies(Collection<EdgeState> discoveredEdges) {
         visitOwners(discoveredEdges);
     }
 
+    private boolean hasNewConstraints() {
+        return upcomingNoLongerPendingConstraints != null && !upcomingNoLongerPendingConstraints.isEmpty();
```

Comment:
:o: I think we could be slightly faster if we only check for `null` here, and in `visitAdditionalConstraints`, we set to `null` instead of clearing.

Reply:
Done